### PR TITLE
BUG: geometry constructors now support `Decimal` inputs

### LINKED ISSUE #1707: Shapely 2.0 doesn't accept Decimal Fields in the Polygon constructor.
## Expected behavior and actual behavior.
Before the Shapely 2.0 update our code worked:

```
    def to_geo(self):
        p1 = (self.Longitude1, self.Latitude1)
        p2 = (self.Longitude2, self.Latitude2)
        p3 = (self.Longitude3, self.Latitude3)
        p4 = (self.Longitude4, self.Latitude4)
        p5 = (self.Longitude1, self.Latitude1)
        poly_coords = (p1, p2, p3, p4, p5)

        poly = Polygon(poly_coords)
        poly = poly.convex_hull

        return poly
```
now it returns a:
`ValueError: Inconsistent coordinate dimensionality`

I've run this also with the latest on master and gotten the same result. We are able to mitigate this by casting the Decimals to floats, but I'd rather not if this is truly a regression and not an intended change.

now it 
## Steps to reproduce the problem.

```
 # the following are latitude/longitude values normally on a Django model, but I've just changed them to the value for reproducibility. 
 from shapely.geometry import Polygon
 from decimal import Decimal
 p1 = (Decimal('-76.572088'), Decimal('39.256796'))
 p2 = (Decimal('-76.569245'), Decimal('39.258841'))
 p3 = (Decimal('-76.567123'), Decimal('39.257046'))
 p4 = (Decimal('-76.571833'), Decimal('39.255359'))
 p5 = (Decimal('-76.572088'), Decimal('39.256796'))
 poly_coords = (p1, p2, p3, p4, p5)
 
 poly = Polygon(poly_coords)
 poly = poly.convex_hull
```

## Operating system

Ubuntu 22.04.1 LTS, but also recreated in a python docker image

## Shapely version and provenance

shapely==2.0.0 installed from PyPI using pip

### PR DESCRIPTION
**Pull request by idanmiara**
Closes #1707

### CODE REVIEW

File: shapely/tests/geometry/test_point.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def test_from_invalid():
         Point(1, 2, 3, 4)
 
     # this worked in shapely 1.x, just ignoring the other coords
-    with pytest.raises(ValueError, match="Invalid values passed"):
+    with pytest.raises(TypeError, match="only size-1 arrays"):
```
Comment (jorisvandenbossche) [inline shapely/tests/geometry/test_point.py:92]: The current error is not super informative, but this is also not an improvement (and also ValueError is better I think). I'll add a comment on the actual code about maybe catching this explicitly and raising a better error message.

File: shapely/geometry/point.py [hunk 2]
```diff
@@ -64,12 +64,12 @@ def __new__(self, *args):
             # Accept either (x, y) or [(x, y)]
             if not hasattr(coords, "__getitem__"):  # generators
                 coords = list(coords)
-
             coords = np.asarray(coords).squeeze()
         else:
             # 2 or 3 args
             coords = np.array(args).squeeze()
 
+        coords = [float(c) for c in coords]
```
Comment (jorisvandenbossche) [inline shapely/geometry/point.py:72]: We could avoid doing this if we already have a numeric numpy array. So maybe something like:

```
if not np.issubdtype(coords.dtype, np.number):
    coords = [float(c) for c in coords]
```

and related to the other comment, we could maybe also first check explicitly for >1d array case to provide a better error message (untested):

```
if coords.ndim > 1:
    raise ValueError("...")
```

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks! Just a some minor points about the Point constructor

File: shapely/geometry/point.py (hunk 2 above)
Comment (idanmiara) [reply shapely/geometry/point.py]: Done.

File: shapely/tests/geometry/test_decimal.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    [(60.0, 80.0, 1.0), (80.0, 80.0, 2.0), (70.0, 60.0, 3.0), (60.0, 80.0, 1.0)],
+    [(30.0, 10.0, 2.0), (50.0, 10.0, 3.0), (40.0, 30.0, 4.0), (30.0, 10.0, 2.0)],
+    [(90.0, 10, 3.0), (110.0, 10.0, 4.0), (100.0, 30.0, 5.0), (90.0, 10.0, 6.0)],
+]
+
+
+@pytest.mark.parametrize("items", [items2d, items3d])
+def test_decimal(items):
+    assert Point(items[0][0]) == Point(to_decimal(items[0][0]))
+    assert Point(*items[0][0]) == Point(*to_decimal(items[0][0]))
```
Comment (sgillies) [inline shapely/tests/geometry/test_decimal.py:58]: @idanmiara every instance of `assert` in this function should be a different test. Please find a way to parameterize or add new `def` statements.

**Review by sgillies — CHANGES_REQUESTED** [review]
I have a major change request with regards to the test and a couple other suggestions.

File: shapely/tests/geometry/test_decimal.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    [(0.0, 0.0, 1.0), (70.0, 120.0, 2.0), (140.0, 0.0, 3.0), (0.0, 0.0, 1.0)],
+    [(60.0, 80.0, 1.0), (80.0, 80.0, 2.0), (70.0, 60.0, 3.0), (60.0, 80.0, 1.0)],
+    [(30.0, 10.0, 2.0), (50.0, 10.0, 3.0), (40.0, 30.0, 4.0), (30.0, 10.0, 2.0)],
+    [(90.0, 10, 3.0), (110.0, 10.0, 4.0), (100.0, 30.0, 5.0), (90.0, 10.0, 6.0)],
+]
+
+
+@pytest.mark.parametrize("items", [items2d, items3d])
+def test_decimal(items):
+    assert Point(items[0][0]) == Point(to_decimal(items[0][0]))
```
Comment (sgillies) [inline shapely/tests/geometry/test_decimal.py:57]: I don't agree with the use of to_decimal here. It's unnecessarily indirect. Why not just give instances of Decimal, explicitly?

File: shapely/geometry/point.py [hunk 5]
```diff
@@ -64,15 +64,19 @@ def __new__(self, *args):
...
             coords = np.asarray(coords).squeeze()
         else:
             # 2 or 3 args
             coords = np.array(args).squeeze()
 
+        if coords.ndim > 1:
+            raise ValueError(
+                "Invalid values passed to Point constructor, "
+                "only size-1 arrays are accepted"
+            )
```
Comment (sgillies) [inline shapely/geometry/point.py:76]: It looks like this code isn't tested. We need a failing test to demonstrate that lines 72-76 are needed.

Comment (idanmiara) [reply shapely/geometry/point.py:76]: It is tested in the original test, see discussion about it above:
https://github.com/shapely/shapely/pull/1720#discussion_r1071623475

File: shapely/tests/geometry/test_decimal.py (hunk 3 above)
Comment (idanmiara) [reply shapely/tests/geometry/test_decimal.py:58]: OK, done.

File: shapely/tests/geometry/test_decimal.py (hunk 4 above)
Comment (idanmiara) [reply shapely/tests/geometry/test_decimal.py:57]: I'm not sure if you were referring to the use of the function `to_decimal` in general in this test or just to the `Point` test.
Anyway, I've parametrized the test as you requested above. 
If you'd prefer to remove `to_decimal` I could hardcode `Decimal` in copies of `items2d` and `items3d` and remove `to_decimal` - otherwise please explain what you meant please.

**Comment by idanmiara** [comment]
> I have a major change request with regards to the test and a couple other suggestions.

No problem. Thanks for your review. Please advise if this is OK now.

File: shapely/geometry/point.py [hunk 6]
```diff
@@ -64,15 +64,19 @@ def __new__(self, *args):
...
-
             coords = np.asarray(coords).squeeze()
         else:
             # 2 or 3 args
             coords = np.array(args).squeeze()
 
+        if coords.ndim > 1:
+            raise ValueError(
+                "Invalid values passed to Point constructor, "
+                "only size-1 arrays are accepted"
```
Comment (jorisvandenbossche) [inline shapely/geometry/point.py:75]: You can remove this second part, as that is exactly the confusing error message from passing an array to the Decimal(..) constructor (which happens if ndim > 1) that we want to avoid raising by checking this manually here. 

Maybe we can append the first part with something like ".. can only pass 2 or 3 coordinates".

Comment (idanmiara) [reply shapely/geometry/point.py:75]: `TypeError(f"Point() takes at most 3 arguments ({len(args)} given)")`  - is raised when more then 3 arguments are given
so maybe in a similar way: `ValueError(f"Point() takes only scalar or 1-size vector arguments, got {args}")`

File: shapely/tests/geometry/test_decimal.py (hunk 4 above)
Comment (sgillies) [reply shapely/tests/geometry/test_decimal.py]: Here's what I've come to understand about using convenience functions in tests. I used to do this somewhat often. But then a year or two later when I wanted to refactor code or update a feature, I would discover that the convenience function wasn't appropriate anymore and actually hindered the refactoring work. This is the awful truth about tests: they can easily make future changes hard. Shapely needs tests, but not too many, and not too fancy.

Please do remove to_decimal.

File: shapely/tests/geometry/test_decimal.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    [(30.0, 10.0), (50.0, 10.0), (40.0, 30.0), (30.0, 10.0)],
+    [(90.0, 10), (110.0, 10.0), (100.0, 30.0), (90.0, 10.0)],
+]
+
+items3d = [
+    [(0.0, 0.0, 1.0), (70.0, 120.0, 2.0), (140.0, 0.0, 3.0), (0.0, 0.0, 1.0)],
+    [(60.0, 80.0, 1.0), (80.0, 80.0, 2.0), (70.0, 60.0, 3.0), (60.0, 80.0, 1.0)],
+    [(30.0, 10.0, 2.0), (50.0, 10.0, 3.0), (40.0, 30.0, 4.0), (30.0, 10.0, 2.0)],
+    [(90.0, 10, 3.0), (110.0, 10.0, 4.0), (100.0, 30.0, 5.0), (90.0, 10.0, 6.0)],
+]
```
Comment (sgillies) [inline shapely/tests/geometry/test_decimal.py:52]: @idanmiara I thought of another very important point!

Having multiple cases is good. However, all of these are basically the same case. Casting all of them to float will succeed because they are all well-formed numbers. We can reduce these lists to a single item and we would have equally good testing. There's a lot of redundancy here and that can bite us in the future if we change anything.

We only really need one 2D item and one 3D item to test the successful cases here.

To test weird edge cases we would want to check decimals that can't be exactly represented by floats (like by having hundreds of decimal places, etc). But that doesn't have to be done in this PR :sweat_smile:

File: shapely/tests/geometry/test_decimal.py (hunk 4 above)
Comment (sgillies) [reply shapely/tests/geometry/test_decimal.py]: This is important work @idanmiara and you've solved the problem well! I want to help you get the tests right.

Comment (idanmiara) [reply shapely/tests/geometry/test_decimal.py]: @sgillies thanks, I see your point. while removing `to_decimal` I've added another test for mixed `decimal` and non-decimal inputs.

File: shapely/tests/geometry/test_decimal.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        items3d_mixed,
+        items3d_decimal,
+    ]
+]
+
+
+@pytest.mark.parametrize("geoms", list(zip(*all_geoms)))
+def test_decimal(geoms):
+    assert geoms[0] == geoms[1] == geoms[2]
+    assert geoms[3] == geoms[4] == geoms[5]
```
Comment (sgillies) [inline shapely/tests/geometry/test_decimal.py:117]: Since we only changed the LinearRing, LineString, and Point constructors my preference would be to have just `test_linearring()`, `test_linestring()`, and `test_point()` within `test_decimal.py`. I think that would keep this test module easy to read and modify over time.

However, I don't insist on that. The work here is good and the changes I am thinking of could be done later.

Thanks @idanmiara !

**Review by sgillies — APPROVED** [review]

File: shapely/tests/geometry/test_decimal.py (hunk 8 above)
Comment (idanmiara) [reply shapely/tests/geometry/test_decimal.py:117]: Thanks for your review @sgillies !

> Since we only changed the LinearRing, LineString, and Point constructors...

I'm included to keep all these tests since implementations sometimes change, and we want to make sure this doesn't break in future refactors.

If you could find the time I'd be happy if you could review my other Shapely opened PRs
Thanks again!

File: shapely/geometry/point.py [hunk 9]
```diff
@@ -64,12 +64,17 @@ def __new__(self, *args):
...
 
+        if coords.ndim > 1:
+            raise ValueError(
+                f"Point() takes only scalar or 1-size vector arguments, got {args}"
+            )
+        if not np.issubdtype(coords.dtype, np.number):
+            coords = [float(c) for c in coords]
         geom = shapely.points(coords)
         if not isinstance(geom, Point):
             raise ValueError("Invalid values passed to Point constructor")
```
Comment (jorisvandenbossche) [inline shapely/geometry/point.py:80]: @idanmiara I added this back. I know this is currently not needed (I don't think a test covers this), and I also can't directly think of a case where `shapely.points(..)` would not return a single Point that wouldn't already be catched by the `if coords.ndim > 1:` check. 
But, Point() really should ever only return a single Point, and `shapely.points()` doesn't guarantee that, so it feels safer to keep this anyway (our tests might not be complete).

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!
